azure-samples/src/main/java/com/microsoft/azure/management/compute/samples/CreateVirtualMachinesAsyncTrackingRelatedResources.java

```diff
@@ -0,0 +1,321 @@
+                //
+                // Resources will be created across multiple threads and the order in which they will be emitted is unpredictable.
+                // Since the SDK and the RX framework handle the multi-threading and parallelization under the hood, our code below
+                // does not need to worry about it, just handle each resource as it is emitted.
+                //
+                // Thanks to how RX works, we can also assume our callback below will be always called on the same thread, so we do
+                // not need to worry about making the collections it uses thread-safe.
+                .map(new Func1<Indexable, Indexable>() {
+                    @Override
+                    public Indexable call(Indexable createdResource) {
+
+                        // Since the resources are of different types, each resources is emitted as the Indexable base type
+                        // so it needs to be cast and handled depending on its type
+                        if (createdResource instanceof Resource) {
+                            Resource resource = (Resource) createdResource;
+
+                            // Report the creation of a resource in the UI
+                            System.out.println(String.format("\tCreated: %s '%s'",
+                                    ResourceUtils.resourceTypeFromResourceId(resource.id()),
+                                    ResourceUtils.nameFromResourceId(resource.id())));
+
+                            if (resource instanceof VirtualMachine) {
+                                // Track the successful creation of virtual machines, so that their related resources do not cleaned up later
+                                VirtualMachine virtualMachine = (VirtualMachine) resource;
+
+                                // Record that this VM was created successfully
+                                vmDefinitions.remove(virtualMachine.key());
+
+                                // Remove the associated resources from cleanup list
+                                vmNonNicResourceDefinitions.remove(virtualMachine.key());
+
+                                // Remove the associated NIC from cleanup list
+                                nicDefinitions.remove(virtualMachine.key());
+                            } else {
+                                // Since this is not a VM, add this resource to the potential cleanup list
+                                createdResourceIds.put(resource.key(), resource.id());
+                            }
+                        }
+                        return createdResource;
+                    }
+               })
+               .onErrorReturn(new Func1<Throwable, Indexable>() {
+                   // Aggregate all the errors
+                   @Override
+                   public Indexable call(Throwable throwable) {
+                       errors.add(throwable);
+                       return null;
+                   }
```

Comment:
Since it is guaranteed to emit only one error (rx.CompositeException or single Throwable) if there is any, we can use `.doOnError()` instead of `onErrorReturn ` and store error in a single error variable as oppose to in a list.

Reply:
Sorry, we still need onErrorReturn due to last() operator